quickwit-cluster/docs/kubernetes-discovery.md

```diff
@@ -0,0 +1,220 @@
+in the DNS record. We do have guarantees that the responses to the different pods will be consistent. Pods will
+restart until one IP is registered, and the same first peer will be returned to all pods, avoiding a split brain.
+
+When adding nodes to the cluster (by scaling up the deployment/statefulset), new nodes will retrieve
+[up to 1000](https://kubernetes.io/docs/concepts/services-networking/service/#over-capacity-endpoints)
+IPs as seeds by querying the service. Existing nodes will not update their seeds, but discover the new nodes
+via the gossip messages.
+
+If an existing node restarts (due to an out-of-memory error for example), it will request up-to-date seeds again
+from the DNS servers, discarding its previous state. This is better than running the DNS enumeration in an
+[init container](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/), which would not
+update the seeds on container restart.
+
+### 2. Gate cluster membership on new a `cluster_name` property
+
+Redis, Cassandra and Elasticsearch all had to solve the IP reuse problem before. They all settled on the simplest and
+most resilient solution: adding a `cluster_name` property to the nodes, and forbidding nodes to join if their
+advertised cluster name does not match the other peers. This mechanism can be implemented in `scuttlebutt` with
+the following changes:
+
+- adding a new required argument `cluster_name` to the `ScuttleServer::spawn` function (String type)
+- copying that value as a new  `cluster_name` field to the `Digest` object
+- updating `ScuttleServer::process_message` to reject messages if it detects a mismatch between the `Digest`'s
+cluster name and its own. A warning will be logged to alert the operators that nodes are misconfigured
+
+Making the cluster name required will simplify the `scuttlebutt` code. To keep single-node deploys easy,
+Quickwit can provide a default value if its `cluster_name` configuration option is not set. Cassandra uses
+["Test Cluster"](https://cassandra.apache.org/doc/latest/cassandra/configuration/cass_yaml_file.html#cluster_name)
+to nudge operators into setting a relevant value, even if they only have one cluster for now.
+
+### 3. Use the pod name as default `node_id`
+
+Pods are assigned a unique name by the system on creation. Reusing this pod name as the `node_id` will make it easier
+for cluster operators to navigate between the Quickwit admin and their infrastructure monitoring. This name is available
+in the `$POD_NAME` environment variable and can be injected in arbitrarily named environment variables using a 
+[fieldRef](https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/).
+
+The best practice for Kubernetes-native services is to allow overriding configuration values with envvar.
+We will update `quickwit-config` to lookup the `QUICKWIT_NODE_ID` envvar, and use its value if not empty.
```

Comment:
This is something that often comes up in discussions.

I wonder if we should offer string interpolation in the quickwit config, using environment variables as a context `${QUICKWIT_NODE_ID}`, possibly with default values `${QUICKWIT_CLUSTER_ID:"test_id"}` maybe a bit like bash does.

I feel like one of the benefit is to leave an explicit trail about the population of a value.
"How did I set the cluster_id? Oh it is a environment value called blabla"

As opposed to having to look at the doc.

Reply:
Adding string interpolation and setting `node_id: $POD_NAME` in the example k8s manifests would be a quick win here, but having automatic envvar mapping would be a cheap way to gain cookie points with seasoned k8s users.

The Datadog agent has its main config as a yaml file, but every option is automatically mapped to the `DD_OPTION_NAME` envvar, which allows to run simple deployments without even creating a configmap to provide the configuration.

Reply:
and there is another option, that Adrien likes, which is override via CLI arguments.
( " ^_^ )

Reply:
For reference:
https://github.com/quickwit-oss/quickwit/issues/1011